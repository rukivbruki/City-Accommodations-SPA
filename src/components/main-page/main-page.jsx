import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {compose} from 'recompose';

import Spinner from '../spinner/spinner.jsx';
import PlacesList from '../places-list/places-list.jsx';
import CitiesList from '../cities-list/cities-list.jsx';
import {ActionCreator as ActiveCityActionCreator} from '../../reducers/active-city/active-city';

import withSortOpen from '../../hocs/with-sort-open/with-sort-open';
import withPlacesSort from '../../hocs/with-places-sort/with-places-sort';
import withActivePlace from '../../hocs/with-active-place/with-active-place';

import {loadOffers} from '../../reducers/offers/offers';

import {getActiveCityOffers, getCities, getIsLoaded, getLoadingError} from '../../reducers/offers/selectors';
import {getActiveCity} from '../../reducers/active-city/selectors';
import {setBodyRootClass} from "../../utils/set-body-root-class";

const PlacesListWrapped = withPlacesSort(withSortOpen(PlacesList));

class MainPage extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    setBodyRootClass(`page page--gray page--main`, `full-height`);
    const {load} = this.props;
    load();
  }

  render() {
    const {
      isLoaded,
      places,
      cities,
      activeCity,
      onCityClick,
      activePlace,
      onActivatePlace,
    } = this.props;

    if (!isLoaded) {
      return <Spinner />;
    }

    return <>
      <main className={`page__main page__main--index
      ${!places.length ? `page__main--index-empty` : ``}`}>
        <h1 className="visually-hidden">Cities</h1>
        <CitiesList
          cities={cities}
          activeCity={activeCity}
          onCityClick={(city) => {
            onActivatePlace(null);
            onCityClick(city);
          }}
        />
        <PlacesListWrapped
          places={places}
          city={activeCity}
          key={activeCity.name}
          activePlace={activePlace}
          onActivatePlace={onActivatePlace}
        />
      </main>
    </>;
  }
}

MainPage.propTypes = {
  places: PropTypes.array,
  cities: PropTypes.array,
  activeCity: PropTypes.object,
  activePlace: PropTypes.object,
  isLoaded: PropTypes.bool.isRequired,
  loadingError: PropTypes.string,
  load: PropTypes.func.isRequired,
  onActivatePlace: PropTypes.func.isRequired,
  onCityClick: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  activeCity: getActiveCity(state),
  places: getActiveCityOffers(state),
  cities: getCities(state),
  isLoaded: getIsLoaded(state),
  loadingError: getLoadingError(state),
});

const mapDispatchToProps = (dispatch) => ({
  onCityClick: (city) => dispatch(ActiveCityActionCreator.setActiveCity(city)),
  load: () => dispatch(loadOffers())
});

export {MainPage};
export default compose(
    connect(mapStateToProps, mapDispatchToProps),
    withActivePlace
)(MainPage);

