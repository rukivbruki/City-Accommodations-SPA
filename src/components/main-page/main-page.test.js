import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import {MainPage} from './main-page.jsx';
import mockPlaces from '../../mocks/mock-offers.js';
import mockCities from '../../mocks/mock-cities.js';

jest.mock(`../../utils/set-body-root-class`);

it(`MainPage snapshot`, () => {
  const renderer = new ShallowRenderer();
  const result = renderer.render(<MainPage
    places={mockPlaces}
    cities={mockCities}
    activeCity={mockCities[0]}
    activePlace={null}
    isLoaded={true}
    loadingError={null}
    load={jest.fn()}
    onCityClick={jest.fn()}
    onActivatePlace={jest.fn()}
  />);
  expect(result).toMatchSnapshot();
});
