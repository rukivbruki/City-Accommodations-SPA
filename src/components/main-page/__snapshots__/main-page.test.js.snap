// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage snapshot 1`] = `
<React.Fragment>
  <main
    className="page__main page__main--index
      "
  >
    <h1
      className="visually-hidden"
    >
      Cities
    </h1>
    <CitiesList
      activeCity={
        Object {
          "location": Object {
            "latitude": 1,
            "longitude": 2,
            "zoom": 13,
          },
          "name": "City name 1",
        }
      }
      cities={
        Array [
          Object {
            "location": Object {
              "latitude": 1,
              "longitude": 2,
              "zoom": 13,
            },
            "name": "City name 1",
          },
          Object {
            "location": Object {
              "latitude": 2,
              "longitude": 3,
              "zoom": 13,
            },
            "name": "City name 2",
          },
        ]
      }
      onCityClick={[Function]}
    />
    <WithPlacesSort
      activePlace={null}
      city={
        Object {
          "location": Object {
            "latitude": 1,
            "longitude": 2,
            "zoom": 13,
          },
          "name": "City name 1",
        }
      }
      onActivatePlace={[MockFunction]}
      places={
        Array [
          Object {
            "bedrooms": 2,
            "city": Object {
              "location": Object {
                "latitude": 1,
                "longitude": 2,
                "zoom": 13,
              },
              "name": "City name 1",
            },
            "description": "text",
            "goods": Array [
              "one",
              "two",
              "tree",
            ],
            "host": Object {
              "avatarUrl": "url",
              "id": 1,
              "isPro": true,
              "name": "name",
            },
            "id": 999,
            "images": Array [
              "url1",
              "url2",
              "url3",
            ],
            "isFavorite": true,
            "isPremium": true,
            "location": Object {
              "latitude": 1,
              "longitude": 2,
              "zoom": 13,
            },
            "maxAdults": 2,
            "picture": "test-image1.jpg",
            "previewImage": "",
            "price": 111,
            "rating": 22,
            "title": "Test title 1",
            "type": "apartment",
          },
          Object {
            "bedrooms": 3,
            "city": Object {
              "location": Object {
                "latitude": 1,
                "longitude": 2,
                "zoom": 13,
              },
              "name": "City name 2",
            },
            "coordinates": Array [
              52.3909553943508,
              4.85309666406198,
            ],
            "description": "text",
            "goods": Array [
              "one",
              "two",
              "tree",
            ],
            "host": Object {
              "avatarUrl": "url",
              "id": 1,
              "isPro": true,
              "name": "name",
            },
            "id": 222,
            "images": Array [
              "url1",
              "url2",
              "url3",
            ],
            "isFavorite": true,
            "isPremium": false,
            "location": Object {
              "latitude": 1,
              "longitude": 2,
              "zoom": 13,
            },
            "maxAdults": 1,
            "picture": "test-image2.jpg",
            "previewImage": "",
            "price": 222,
            "rating": 100,
            "title": "Test title 2",
            "type": "house",
          },
        ]
      }
    />
  </main>
</React.Fragment>
`;
